### PR TITLE
Add support for most standard `.editorconfig` settings

### PR DESCRIPTION
**Pull request by Colengms**
Addresses: https://github.com/microsoft/vscode-cpptools/issues/7920

Adds support for:
* `indent_style`
* `indent_size`
* `tab_size`
* `end_of_line` (applied any time the document is activated, as an unsaved edit.  This is equivalent to selecting the newline mode from the menu at the bottom of the window).
* `insert_final_newline` (applied only when "Format Document" is used, so also on save if using `editor.formatOnSave`.)

vcFormat appears to always do `trim_trailing_whitespace`.

That leaves `charset`.  Support for which appears to be blocked on: https://github.com/microsoft/vscode/issues/824

### CODE REVIEW

File: Extension/src/LanguageServer/client.ts [hunk 1]
```diff
@@ -2387,10 +2390,47 @@ export class DefaultClient implements Client {
         return this.requestWhenReady(() => this.languageClient.sendRequest(SwitchHeaderSourceRequest, params));
     }
 
+    private async updateActiveDocumentTextOptions(): Promise<void> {
+        const editor: vscode.TextEditor | undefined = vscode.window.activeTextEditor;
+        if (editor
+            && editor.document
+            && editor.document.uri.scheme === "file"
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/client.ts]: You could remove `editor && editor.document` if you use `editor?.document?.uri.scheme`

File: Extension/src/LanguageServer/Providers/documentFormattingEditProvider.ts [hunk 2]
```diff
@@ -42,6 +42,31 @@ export class DocumentFormattingEditProvider implements vscode.DocumentFormatting
                     newText: textEdit.newText
                 });
             });
+            // Apply insert_final_newline from .editorconfig
+            if (document.lineCount > 0 && editorConfigSettings !== undefined && editorConfigSettings.insert_final_newline) {
+                // Check if there is already a newline at the end.  If so, formatting edits should not replace it.
+                if (!document.lineAt(document.lineCount - 1).isEmptyOrWhitespace) {
+                    const endPosition: vscode.Position = document.lineAt(document.lineCount - 1).range.end;
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/Providers/documentFormattingEditProvider.ts]: You could avoid this 2nd call if you saved the 1st call in a variable.

File: Extension/src/LanguageServer/client.ts [hunk 3]
```diff
@@ -2387,10 +2390,47 @@ export class DefaultClient implements Client {
...
+            && (editor.document.languageId === "c"
+                || editor.document.languageId === "cpp"
+                || editor.document.languageId === "cuda-cpp")) {
+            // If using vcFormat, check for a ".editorconfig" file, and apply those text options to the active document.
+            const settings: CppSettings = new CppSettings();
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/client.ts]: I believe you should be passing the rootUri here or else you might not get the settings for the current workspace folder.

**Review by sean-mcmanus — CHANGES_REQUESTED** [review]

File: Extension/src/LanguageServer/client.ts (hunk 3 above)
Comment (Colengms) [reply Extension/src/LanguageServer/client.ts]: Hmm.  I'm wondering if this is correct in a few other places.  Should the existing occurrence in `onDidChangeSettings()` also use `this.RootUri`?

**Review by sean-mcmanus — APPROVED** [review]

File: Extension/src/LanguageServer/client.ts (hunk 3 above)
Comment (sean-mcmanus) [reply Extension/src/LanguageServer/client.ts]: It's incorrect whenever a "resource" or "machine-overridable" scope setting is accessed, which is the case for "formatting" in onDidChangeSettings -- we could have other cases too.

**Pull request merged**